### PR TITLE
Update prefix utilization calculation

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes: #2916 #3894
# What's Changed

- If the `Prefix.type` is `Container`, the utilization is calculated as the sum of the total address space of all child prefixes.
- If the `Prefix.type` is `Pool`, the utilization is calculated as the sum of the total number of IP addresses within the pool's range.
- If the `Prefix.type` is `Network`:
    - The utilization is calculated as the sum of the total address space of all child `Pool` prefixes plus the total number of child IP addresses.
    - For IPv4 networks larger than /31, if neither the first or last address is occupied by either a pool or an IP address, they are subtracted from the total size of the prefix.

## Example

- 192.168.0.0/16          `Container - 1024/65536 utilization`
    - 192.168.1.0/24      `Network - 1/254 utilization`
        - 192.168.1.1     `IP Address`
    - 192.168.2.0/24      `Network - 4/256 utilization`
        - 192.168.2.0/30  `Pool - 1/4 utilization`
            - 192.168.2.1 `IP Address`
    - 192.168.3.0/24      `Network - 5/254 utilization`
        - 192.168.3.1     `IP Address`
        - 192.168.3.64/30 `Pool - 0/4 utilization`
    - 192.168.4.0/24      `Network - 1/256 utilization`
        - 192.168.4.255   `IP Address`

## ⚠️ NOTE

It's no longer possible to assign IP addresses to a pool and pools must exist within a network with a shorter prefix length. So in order to create a /24 pool of 256 addresses, you must first create a network to contain it that is /23 or larger.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/ipam/models.py [hunk 1]
```diff
@@ -709,7 +709,6 @@ def subnets(self, direct=False, include_self=False, for_update=False):
 
         return query.filter(
             ip_version=self.ip_version,
-            prefix_length__gte=self.prefix_length,
```
Comment (gsnider2195) [inline nautobot/ipam/models.py:712]: With uniqueness constraints it would never be possible to have an overlapping network with the same prefix length and if the prefix length is shorter, either `network__gte` or `broadcast__lte` would not match.

File: nautobot/docs/installation/upgrading-from-nautobot-v1.md [hunk 2]
```diff
@@ -91,14 +91,38 @@ The following changes have been made to the `Prefix` model.
...
-#### IPAddressd Parenting Concrete Relationship
+### IPAddress Parenting Concrete Relationship
 
 The `ipam.IPAddress` model has been modified to have a foreign key to `ipam.Prefix` as the `parent` field. Parenting of IP addresses is now automatically managed at the database level to greatly improve performance especially when calculating tree hierarchy and utilization.
 
 | Removed                | Replaced With   |
 |------------------------|-----------------|
 | `get_child_prefixes()` | `descendants()` |
 
+### Prefix get_utilization Method
```
Comment (glennmatthews) [inline nautobot/docs/installation/upgrading-from-nautobot-v1.md:102]: Would it make sense to duplicate some/all of this section into `nautobot/docs/models/ipam/prefix.md`?

File: nautobot/ipam/models.py [hunk 3]
```diff
@@ -836,29 +835,42 @@ def get_first_available_ip(self):
...
+            child_ips = netaddr.IPSet(self.ip_addresses.values_list("host", flat=True))
+
+        child_prefixes = netaddr.IPSet(p.prefix for p in self.children.only("network", "prefix_length").iterator())
+        numerator = child_prefixes | child_ips
+
+        # Exclude network and broadcast address from the denominator unless they've been assigned to an IPAddress or child pool.
+        if all(
+            [
+                not self.children.exclude(type=choices.PrefixTypeChoices.TYPE_POOL).exists(),
+                denominator > 2,
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:865]: This line is for /31, /32 prefixes? A comment might be good to document this.

File: nautobot/ipam/models.py [hunk 4]
```diff
@@ -836,29 +835,42 @@ def get_first_available_ip(self):
...
+                self.type != choices.PrefixTypeChoices.TYPE_POOL,
+                self.ip_version == 4,
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:867]: I saw your comment on Slack about the behavior of `all()`, but can we make this any more efficient by doing the non-DB-accessing checks first? I.e not even look at `self.children` if we don't pass the other three checks?

File: nautobot/ipam/models.py [hunk 5]
```diff
@@ -836,29 +835,42 @@ def get_first_available_ip(self):
...
-                    self.type != choices.PrefixTypeChoices.TYPE_POOL,
-                ]
-            ):
-                prefix_size -= 2
-            child_count = prefix_size - self.get_available_ips().size
-            return UtilizationData(numerator=child_count, denominator=prefix_size)
+            child_ips = netaddr.IPSet(self.ip_addresses.values_list("host", flat=True))
+
+        child_prefixes = netaddr.IPSet(p.prefix for p in self.children.only("network", "prefix_length").iterator())
+        numerator = child_prefixes | child_ips
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:859]: Perhaps `numerator_set` for clarity?

File: nautobot/ipam/models.py [hunk 6]
```diff
@@ -836,29 +835,42 @@ def get_first_available_ip(self):
         return f"{next(available_ips.__iter__())}/{self.prefix_length}"
 
     def get_utilization(self):
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:837]: Does this help at all with #2248?

File: nautobot/ipam/tables.py [hunk 7]
```diff
@@ -410,8 +410,8 @@ class PrefixDetailTable(PrefixTable):
     class Meta(PrefixTable.Meta):
         fields = (
             "pk",
-            "namespace",
             "prefix",
+            "namespace",
```
Comment (glennmatthews) [inline nautobot/ipam/tables.py:414]: Was there an open issue regarding this ordering? I thought there was but I can't find it.

File: nautobot/ipam/tests/test_models.py [hunk 8]
```diff
@@ -522,57 +522,53 @@ def test_get_utilization(self):
...
+            prefix="10.0.0.128/30", type=PrefixTypeChoices.TYPE_POOL, status=self.status, namespace=self.namespace
+        )
+        self.assertEqual(slash25.get_utilization(), (4, 128))
+
+        # When the pool does not overlap with broadcast or network address, the denominator decrements by 2
+        pool.network = "10.0.0.132"
+        pool.save()
+        self.assertEqual(slash25.get_utilization(), (4, 126))
 
         # IPv4 Non-container Prefix /31, network and broadcast addresses count toward utilization
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_models.py:557]: Can we add a test for /32 and /128 prefixes, just making sure we aren't setting ourselves up for a divide-by-zero or anything like that?

File: nautobot/ipam/tests/test_models.py [hunk 9]
```diff
@@ -581,13 +577,10 @@ def test_get_utilization(self):
...
-            Prefix(prefix="22.0.0.0/12", status=self.status, namespace=self.namespace),
-            Prefix(prefix="22.16.0.0/12", status=self.status, namespace=self.namespace),
-            Prefix(prefix="22.32.0.0/12", status=self.status, namespace=self.namespace),
-            Prefix(prefix="22.48.0.0/12", status=self.status, namespace=self.namespace),
-        ]
-        [p.save() for p in prefixes]  # pylint: disable=expression-not-assigned
+        Prefix.objects.create(prefix="22.0.0.0/12", status=self.status, namespace=self.namespace)
+        Prefix.objects.create(prefix="22.16.0.0/12", status=self.status, namespace=self.namespace)
+        Prefix.objects.create(prefix="22.32.0.0/12", status=self.status, namespace=self.namespace)
+        Prefix.objects.create(prefix="22.48.0.0/12", status=self.status, namespace=self.namespace)
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_models.py:583]: IIRC the reason we were doing `save()` instead of `create()` was because the logic to decompose `prefix` into its components wasn't happening in some cases - is that a non-issue here?

File: nautobot/ipam/tables.py (hunk 7 above)
Comment (gsnider2195) [reply nautobot/ipam/tables.py:414]: I did too and I also can't find it.

File: nautobot/ipam/tests/test_models.py (hunk 9 above)
Comment (gsnider2195) [reply nautobot/ipam/tests/test_models.py:583]: It looks like it's doing almost the same thing by using create:

https://github.com/django/django/blob/722e9f8a38f5b34f2423059a75f8a59bb8eb931a/django/db/models/query.py#L446-L454
```py
    def create(self, **kwargs):
        """
        Create a new object with the given kwargs, saving it to the database
        and returning the created object.
        """
        obj = self.model(**kwargs)
        self._for_write = True
        obj.save(force_insert=True, using=self.db)
        return obj
```

File: nautobot/ipam/tables.py (hunk 7 above)
Comment (gsnider2195) [reply nautobot/ipam/tables.py:414]: Got it now #3894

File: nautobot/ipam/models.py (hunk 6 above)
Comment (gsnider2195) [reply nautobot/ipam/models.py:837]: - With this utilization calculation in place it takes 94 seconds to load the list view with 1000 prefixes. 
- If I short circuit the utilization calculation and just return `UtilizationData(numerator=numerator_set.size, denominator=denominator)` it takes 47 seconds. 
- With the jinja template from #2248 and short circuiting get_utilization it takes 46 seconds.
- With the jinja template and calculating utilization it takes 88 seconds
- When I switch back to `next` it takes 71 seconds

These times are with DJDT enabled

File: nautobot/ipam/querysets.py [hunk 10]
```diff
@@ -327,52 +327,38 @@ def _validate_cidr(self, value):
         except netaddr.AddrFormatError as err:
             raise ValidationError({"cidr": f"{value} does not appear to be an IPv4 or IPv6 network."}) from err
 
-    def get_closest_parent(self, cidr, namespace, max_prefix_length=0):
```
Comment (gsnider2195) [inline nautobot/ipam/querysets.py:330]: Refactored `get_closest_parent` with @jathanism to fix case where /32 IPv4 address was not matching /32 prefix

File: nautobot/ipam/models.py (hunk 6 above)
Comment (glennmatthews) [reply nautobot/ipam/models.py:837]: Not entirely sure I follow all that. :-) Sounds like the gist is that the jinja template gives only a negligible performance gain at this point?

File: nautobot/ipam/tests/test_models.py
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/ipam/tests/test_models.py:1]: I brought the changes for `test_querysets.py` over from [u/gas-prefix-ip-constraints](https://github.com/nautobot/nautobot/tree/u/gas-prefix-ip-constraints). The test in `test_models` didn't need to be updated.

File: nautobot/ipam/models.py (hunk 6 above)
Comment (gsnider2195) [reply nautobot/ipam/models.py:837]: Bottom line is this change makes the utilization calculation slower and it doesn't appear that #2248 is going to help much.

File: nautobot/ipam/models.py [hunk 11]
```diff
@@ -832,29 +831,43 @@ def get_first_available_ip(self):
...
-                    self.prefix_length < 31,
-                    self.type != choices.PrefixTypeChoices.TYPE_POOL,
-                ]
-            ):
-                prefix_size -= 2
-            child_count = prefix_size - self.get_available_ips().size
-            return UtilizationData(numerator=child_count, denominator=prefix_size)
+            child_ips = netaddr.IPSet(self.ip_addresses.values_list("host", flat=True))
+
+        child_prefixes = netaddr.IPSet(p.prefix for p in self.children.only("network", "prefix_length").iterator())
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:854]: Could skip this lookup if self.type != TYPE_CONTAINER, right?

File: nautobot/ipam/models.py [hunk 12]
```diff
@@ -832,29 +831,43 @@ def get_first_available_ip(self):
...
-                [
-                    self.ip_version == 4,
-                    self.prefix_length < 31,
-                    self.type != choices.PrefixTypeChoices.TYPE_POOL,
-                ]
-            ):
-                prefix_size -= 2
-            child_count = prefix_size - self.get_available_ips().size
-            return UtilizationData(numerator=child_count, denominator=prefix_size)
+            child_ips = netaddr.IPSet(self.ip_addresses.values_list("host", flat=True))
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:852]: Could skip this lookup if self.type == TYPE_CONTAINER, right?

**Review by glennmatthews — APPROVED** [review]

**Review by jathanism — APPROVED** [review]
